drv/cosmo-seq-server/src/main.rs

```diff
@@ -0,0 +1,595 @@
+        // Drop the device into reset and hold it there
+        sys.gpio_reset(config.creset);
+        hl::sleep_for(1);
+        let _ = dev.release();
+
+        return Err(SeqError::AuxChecksumMismatch);
+    }
+
+    ringbuf_entry!(Trace::WaitForDone);
+    ice40::finish_bitstream_load(dev, sys, config).map_err(SeqError::Ice40)?;
+    ringbuf_entry!(Trace::Programmed);
+
+    // Bring the user design out of reset
+    sys.gpio_set(SP_TO_FPGA2_SYSTEM_RESET_L);
+
+    Ok(())
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+#[allow(unused)]
+struct ServerImpl {
+    state: PowerState,
+    jefe: Jefe,
+    sys: Sys,
+    hf: HostFlash,
+    seq: fmc_periph::Sequencer,
+}
+
+impl ServerImpl {
+    fn new(token: drv_spartan7_loader_api::Spartan7Token) -> Self {
+        let now = sys_get_timer().now;
+        let seq = fmc_periph::Sequencer::new(token);
+        ringbuf_entry!(Trace::Startup {
+            early_power_rdbks: (&seq.early_power_rdbks).into(),
+        });
+        ringbuf_entry!(Trace::SetState {
+            prev: None, // dummy value
+            next: PowerState::A2,
+            why: StateChangeReason::InitialPowerOn,
+            now,
+        });
+        let jefe = Jefe::from(JEFE.get_task_id());
+        jefe.set_state(PowerState::A2 as u32);
+
+        ServerImpl {
+            state: PowerState::A2,
+            jefe,
+            sys: Sys::from(SYS.get_task_id()),
+            hf: HostFlash::from(HF.get_task_id()),
+            seq,
+        }
+    }
+
+    fn get_state_impl(&self) -> PowerState {
+        self.state
+    }
+
+    /// Logs a set of state registers, returning the state machine states
+    fn log_state_registers(&self) -> StateMachineStates {
+        let seq_api_status = (&self.seq.seq_api_status).into();
+        let nic_api_status = (&self.seq.nic_api_status).into();
+        ringbuf_entry!(Trace::RegStateValues {
+            seq_api_status,
+            seq_raw_status: (&self.seq.seq_raw_status).into(),
+            nic_api_status,
+            nic_raw_status: (&self.seq.nic_raw_status).into(),
+        });
+        StateMachineStates {
+            seq: seq_api_status.a0_sm,
+            nic: nic_api_status.nic_sm,
+        }
+    }
+
+    /// Logs a set of power good registers
+    fn log_pg_registers(&self) {
+        ringbuf_entry!(Trace::RegPgValues {
+            rail_pgs: (&self.seq.rail_pgs).into(),
+            rail_pgs_max_hold: (&self.seq.rail_pgs_max_hold).into(),
+        });
+    }
+
+    fn set_state_impl(
+        &mut self,
+        state: PowerState,
+        why: StateChangeReason,
+    ) -> Result<(), CpuSeqError> {
+        let now = sys_get_timer().now;
+        ringbuf_entry!(Trace::SetState {
+            prev: Some(self.state),
+            next: state,
+            why,
+            now,
+        });
+
+        use fmc_periph::A0Sm;
+        match (self.get_state_impl(), state) {
+            (PowerState::A2, PowerState::A0) => {
```

Comment:
Did we want to add the `CPU_PRESENT` `CORETYPE` `SP5` checks before merging? e.g. https://github.com/oxidecomputer/hubris/blob/master/drv/gimlet-seq-server/src/main.rs#L763

Reply:
Opened #2051 in the interest of getting this merged